include/boost/geometry/algorithms/line_interpolate_point.hpp

```diff
@@ -0,0 +1,402 @@
+        range::push_back(range, p2);
+    }
+};
+
+struct convert_and_assign
+{
+    template <typename Point1, typename Point2>
+    inline void apply(Point1& p1, Point2& p2)
+    {
+        geometry::detail::conversion::convert_point_to_point(p1, p2);
+    }
+
+};
+
+
+/*!
+\brief Internal, calculates interpolation point of a linestring using iterator pairs and
+    specified strategy
+*/
+template <typename Policy>
+struct range
+{
+    template <typename Range, typename PointType, typename Strategy>
+    static inline void apply(Range const& range,
+                             double const& fraction,
+                             PointType & p_or_mp,
+                             Strategy const& strategy)
+    {
+        Policy policy;
+
+        typedef typename boost::range_iterator<Range const>::type iterator_t;
+        typedef typename boost::range_value<Range const>::type point_t;
+
+        typedef typename select_most_precise
+            <
+                typename default_length_result<Range>::type,
+                double
+            >::type calc_t;
+
+        typedef typename Strategy::template result_type<point_t> result_type;
+        typedef boost::container::vector<result_type> vector_t;
+        typedef typename boost::range_iterator<vector_t const>::type viterator_t;
+
+        iterator_t it = boost::begin(range);
+        iterator_t end = boost::end(range);
+
+        if (it == end) // empty(range)
+        {
+            return;
+        }
+        if (fraction == 0)
+        {
+            policy.apply(*it, p_or_mp);
+            return;
+        }
+        if (fraction == 1)
+        {
+            policy.apply(*(end-1), p_or_mp);
+            return;
+        }
+            
+        // compute lengths of segments and their sum
+        calc_t tot_len = 0;
+        vector_t lengths;
+        iterator_t prev = it++;
+        do {
+            result_type len = strategy.compute(*prev, *it);
+            lengths.push_back(len);
+            tot_len += len.distance;
+            prev = it++;
+        } while (it != end);
+
+        //compute interpolation point(s)
+        calc_t repeated_fraction = fraction;
+        calc_t prev_fraction = 0;
+        calc_t cur_fraction = 0;
+        unsigned int i = 1;
+        viterator_t vit = boost::begin(lengths);
+        it = boost::begin(range);
+        prev = it++;
+        bool single_point = false;
+
+        do {
+            calc_t seg_fraction = vit->distance / tot_len;
+
+            cur_fraction = (it + 1 == end) ? 1 : prev_fraction + seg_fraction;
+
+            while (cur_fraction >= repeated_fraction && !single_point)
+            {
+                point_t p;
+
+                strategy.apply(*prev, *it,
+                               (repeated_fraction - prev_fraction) / seg_fraction,
+                               p, *vit);
+                single_point = boost::is_same<Policy, convert_and_assign>::value;
```

Comment:
It is known at compile-time. Why not handle point and multipoint in separate functions or even have `if`-condition before?

Reply:
I removed the `bool` variable and changed the condition to be more understandable here: https://github.com/boostorg/geometry/pull/526/commits/a2e355a683d7aeb64ada41c50e604342ffe69209 
I do not use separate functions for compactness, but I do not know what you have in mind.

Reply:
I will check the new version but to clarify what I had in mind. One function/struct for point calling the strategy once and a separate function for multipoint with a loop. This way it would be clear what's happening in which case.

There is also one additional issue here, i.e. the behavior is based on the Policy type. Why not on the geometry tag? This would be more clear too. Or am I missing something?

But anyway, I'll check the new version.